content/en/js3/blocks/then/index.md

```diff
@@ -0,0 +1,56 @@
++++
+title = '🪆 .then()'
+headless = true
+time = 30
+facilitation = false
+emoji= '🧩'
+[objectives]
+    1='Chain then() on to a Promise'
++++
+
+```mermaid
+graph LR
+    Promise{{🤝 Promises}} --> |resolve to a| Response{📤 Response}
+    Response ---> |sequence with| then{{🪆️ then}}
+```
+
+[`.then()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then) is a method that belongs to the Promise {{<tooltip title="prototype">}}A prototype object is like a template. `then` is a method available on any Promise.{{</tooltip>}} You can think of the commands as
+
+1. _given_ a request to `fetch` some data
+2. _when_ the `response` comes back / the promise resolves to a [response object](https://developer.mozilla.org/en-US/docs/Web/API/Response)
+3. `then` do this next thing with the data / execute this callback
+
+The `.then()` method takes in a [callback](https://www.w3schools.com/js/js_callback.asp) function that will run once the promise resolves.
+
+For example:
+
+```js
+const url = "https://api.github.com/users/SallyMcGrath";
+fetch(url).then((response) => {
+  return response.json(); // <== .json() is the callback (a static method of the response interface)
+});
+```
+
+It's a similar idea as the event loop we have already investigated, but this time we can control it clearly. The .then() method _queues_ up callback functions to execute in sequence once the asynchronous operation completes successfully. This allows us to write code as if it was happening in time order.
+
+{{<note type="tip">}}
+A `Promise` always returns a new `Promise`.
+{{</note>}}
+
+We can chain multiple .then() calls to run more logic, passing the resolved value to the next callback in the chain. This allows us to handle the asynchronous response in distinct steps. Let's create a getProfile function so your Node REPL will execute this chain correctly:
+
+```js
+const getProfile = (url) => {
+  return fetch(url)
+    .then((response) => response.json())
+    .then((data) => data.html_url)
+    .then((htmlUrl) => console.log(htmlUrl));
+};
+getProfile("https://api.github.com/users/SallyMcGrath");
+```
+
+So `then` returns a _new_ `Promise`, and you can call `then` again on the new object. You can chain Promises in ever more complex dependent steps. This is called [Promise chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#chaining).
+
+It's important to understand some of what is happening with Promises and `then`. But for the most part, you will not be writing code in this style.
+
+You will be writing ([declaring](/guides/paradigm/)) a description of what you want to happen and a [loosely-coupled](https://en.wikipedia.org/wiki/Loose_coupling) set of stateless, single-purpose functions to make it so. To help you stay thinking in this way, let's explore async/await.
```

Comment:
This line feels a bit abstract and not very explained... I think we should either expand it or remove it - which do you prefer?

Reply:
Yeah, I agree.

Here's my thinking here: at about this point I often get into long discussions with trainees about why we want them to use map, why we keep harassing them to break up their functions, etc. They don't get the point of it and they feel annoyed by it, like we're deliberately making things harder just to vex them. My best response to this has generally been to explain we're learning a style of programming, that they can go read about if they like. And that there are other styles and that's fine, but we're doing this for now. JS3 is IME when these points reach a crescendo, so I'd like to address it. What do you think? How could you rephrase?

Reply:
I'm inclined to say: Let's remove this from the prep, and try to insert a 30 minute workshop into the day to actively have the discussion (which we can maybe try to put together in the future, through some discussion, rather than trying to make Right Now)?

I think this is something that it's hard to really understand the value of without working through examples (and I'm slightly skeptical that with the current course content we can even motivate that very well - really it's more long-term projects, approaching others' code, and testability where you start seeing the values of this)...

Let's maybe discuss synchronously how these discussions tend to go?